File: app/src/main/java/com/ikvych/cocktail/viewmodel/MainActivityViewModel.java
package com.ikvych.cocktail.viewmodel;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;

import com.ikvych.cocktail.data.entity.Drink;
import com.ikvych.cocktail.data.repository.DrinkDbRepositoryImpl;
import com.ikvych.cocktail.data.repository.base.DrinkDbRepository;
import com.ikvych.cocktail.viewmodel.base.BaseViewModel;

import java.util.Collections;
import java.util.List;

public class MainActivityViewModel extends BaseViewModel {

    private DrinkDbRepository drinkDbRepository;
    private LiveData<List<Drink>> drinks;

    public MainActivityViewModel(@NonNull Application application) {
        super(application);
        drinkDbRepository = new DrinkDbRepositoryImpl(application);
        drinks = drinkDbRepository.getDrinks();
    }

    @Override
    public LiveData<List<Drink>> getDrinksLiveData() {
        return drinks;
    }

    public void saveDrink(Drink drink) {
        drinkDbRepository.saveDrink(drink);
    }

    @Override
    public List<Drink> getCurrentData() {
        if (drinks.getValue() != null) {
            return drinks.getValue();
        }
        return Collections.emptyList();
    }
}

File: app/src/main/java/com/ikvych/cocktail/viewmodel/MainActivityViewModel.kt
package com.ikvych.cocktail.viewmodel

import android.app.Application
import android.content.Context
import android.content.SharedPreferences
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ProcessLifecycleOwner
import androidx.lifecycle.SavedStateHandle
import com.ikvych.cocktail.data.repository.source.CocktailRepository
import com.ikvych.cocktail.listener.ApplicationLifeCycleObserver
import com.ikvych.cocktail.presentation.mapper.CocktailModelMapper
import com.ikvych.cocktail.presentation.model.cocktail.CocktailModel
import java.text.SimpleDateFormat
import java.util.*

class MainActivityViewModel(
    private val cocktailRepository: CocktailRepository,
    private val mapper: CocktailModelMapper,
    application: Application,
    savedStateHandle: SavedStateHandle
) : DrinkViewModel(application, savedStateHandle, cocktailRepository, mapper),
    ApplicationLifeCycleObserver.OnLifecycleObserverListener {

    val cocktailOfTheDayLiveData: MutableLiveData<CocktailModel?> = MutableLiveData()
    private var lifecycleObserver: ApplicationLifeCycleObserver
    private var sharedPreferences: SharedPreferences = application.getSharedPreferences(
        MAIN_ACTIVITY_SHARED_PREFERENCE,
        Context.MODE_PRIVATE
    )
    val navBarTitleVisibilityLiveData: MutableLiveData<Boolean> =
        appSettingRepository.showNavigationBarTitleLiveData
    val showBatteryStateLiveData: MutableLiveData<Boolean> =
        appSettingRepository.showBatteryStateLiveData


    init {
        lifecycleObserver = ApplicationLifeCycleObserver(this, sharedPreferences)
        ProcessLifecycleOwner.get().lifecycle.addObserver(lifecycleObserver)
    }

    override fun onCleared() {
        ProcessLifecycleOwner.get().lifecycle.removeObserver(lifecycleObserver)
        super.onCleared()
    }

    override fun shouldShowDrinkOfTheDay() {
        showDrinkOfTheDay()
    }

    private fun showDrinkOfTheDay() {

        val currentDate = Date()
        val pattern = "MM-dd-yyyy"
        val simpleDateFormat = SimpleDateFormat(pattern, Locale.getDefault())
        val stringDate: String = simpleDateFormat.format(currentDate)

        launchRequest(cocktailOfTheDayLiveData) {
            val repo = cocktailRepository
            val cocktailOfTheDay = repo.findCocktailOfTheDay(stringDate)
            if (cocktailOfTheDay == null) {
                val allDrinks = cocktailRepository.findAllCocktails()
                if (allDrinks.isNullOrEmpty()) {
                    return@launchRequest null
                } else {
                    val newDrinkOfTheDay = allDrinks.random()
                    newDrinkOfTheDay.cocktailOfTheDay = stringDate
                    cocktailRepository.addOrReplaceCocktail(newDrinkOfTheDay)
                    mapper.mapTo(cocktailRepository.findCocktailOfTheDay(stringDate)!!)
                }
            } else {
                mapper.mapTo(cocktailOfTheDay)
            }
        }
    }

    companion object {
        const val MAIN_ACTIVITY_SHARED_PREFERENCE = "MAIN_ACTIVITY_SHARED_PREFERENCE"
    }
}